sockeye/training.py

```diff
@@ -228,40 +275,141 @@ def fit(self,
+
+        if horovod_mpi.using_horovod():
+            # Synchronize shard order across workers
+            steps = horovod_mpi.MPI.COMM_WORLD.bcast(steps, root=0)
+
+        steps_iter = itertools.cycle(steps)
+
         checkpoint_up_to_date = False
+        did_grad_step = False
         while True:
             if self.config.max_epochs is not None and self.state.epoch == self.config.max_epochs:
                 logger.info("Maximum # of epochs (%s) reached.", self.config.max_epochs)
                 if not checkpoint_up_to_date:
                     time_cost = time.time() - tic
-                    self._create_checkpoint(checkpoint_decoder, time_cost, train_iter, validation_iter)
+                    self._create_checkpoint(checkpoint_decoder, loss_prefixes, mt_steps,
+                                            source_lang, target_lang, time_cost, train_iter,
+                                            validation_iter)
                 break
 
             if self.config.max_updates is not None and self.state.updates == self.config.max_updates:
                 logger.info("Maximum # of updates (%s) reached.", self.config.max_updates)
                 if not checkpoint_up_to_date:
                     time_cost = time.time() - tic
-                    self._create_checkpoint(checkpoint_decoder, time_cost, train_iter, validation_iter)
+                    self._create_checkpoint(checkpoint_decoder, loss_prefixes, mt_steps,
+                                            source_lang, target_lang, time_cost, train_iter,
+                                            validation_iter)
                 break
 
             if self.config.max_samples is not None and self.state.samples >= self.config.max_samples:
                 logger.info("Maximum # of samples (%s) reached", self.config.max_samples)
                 if not checkpoint_up_to_date:
                     time_cost = time.time() - tic
-                    self._create_checkpoint(checkpoint_decoder, time_cost, train_iter, validation_iter)
+                    self._create_checkpoint(checkpoint_decoder, loss_prefixes, mt_steps,
+                                            source_lang, target_lang, time_cost, train_iter,
+                                            validation_iter)
                 break
 
-            did_grad_step = self._step(batch=train_iter.next())
-            checkpoint_up_to_date = checkpoint_up_to_date and not did_grad_step
+            step, step_variant = next(steps_iter)
+            if step == "MT":
+                loss_prefix = get_loss_prefix("MT", step_variant[0].name, step_variant[1].name)
+                if step_variant == (source_lang, target_lang):
+                    did_grad_step = self._mt_step(batch=train_iter.next(), loss_prefix=loss_prefix)
+                elif step_variant == (target_lang, source_lang):
+                    did_grad_step = self._mt_step(batch=train_iter.next().reverse(), loss_prefix=loss_prefix)
```

Comment:
What is the effect of `reverse()` here?

Reply:
Reverse takes a source -> target batch and reverses it to target -> source. This is for bi-directional model training.